src/Http/Routing/src/Matching/HttpMethodMatcherPolicy.cs

```diff
@@ -269,12 +266,21 @@ public PolicyJumpTable BuildJumpTable(int exitDestination, IReadOnlyList<PolicyJ
 
         private Endpoint CreateRejectionEndpoint(IEnumerable<string> httpMethods)
         {
+            const string AllowHeaderName = "Allow";
             var allow = string.Join(", ", httpMethods);
+
             return new Endpoint(
                 (context) =>
                 {
                     context.Response.StatusCode = 405;
-                    context.Response.Headers.Add("Allow", allow);
+
+                    // Prevent ArgumentException from duplicate key if header already added, such as when the
+                    // request is re-executed by an error handler (see https://github.com/aspnet/AspNetCore/issues/6415)
+                    if (!context.Response.Headers.ContainsKey(AllowHeaderName))
```

Comment:
`context.Response.Headers[AllowHeaderName] = allow;`

Reply:
Updated. I initially avoided this in case it ended up appending the headers so they doubled-up, but the test I added still passes with it that way.

Reply:
Always setting here is more correct. This code has no way of know if the header had been set to a different value.